packages/devtools/lib/src/timeline/flame_chart_canvas.dart

```diff
@@ -285,6 +293,13 @@ class FlameChartCanvas extends FlameChart {
     for (FlameChartNode node in row.nodes) {
       if (node.rect.left + node.rect.width < visible.left) continue;
       if (node.rect.left > visible.right) break;
+
+      // Do not paint native frames if 'Hide native samples' is checked.
+      if (shouldHideNativeSamples &&
+          node.stackFrame.hasAncestorWithId(CpuProfileData.nativeRootId)) {
```

Comment:
we need to keep this inner loop fast. Now you are traversing up possibly 120 ancestors for every single node in the tree. Propagate this data down the free for efficiency.